client/state/plugins/premium/reducer.js

```diff
@@ -66,20 +65,26 @@ export const plugins = withSchemaValidation( pluginInstructionSchema, ( state =
 				} );
 			}
 			return state;
-		case SERIALIZE:
-			// Save the error state as a string message.
-			return mapValues( state, ( pluginList ) =>
+		default:
+			return state;
+	}
+};
+
+export const plugins = withSchemaValidation(
+	pluginInstructionSchema,
+	withPersistence( pluginsReducer, {
+		// Save the error state as a string message and omit the `key` field.
+		serialize: ( state ) =>
+			mapValues( state, ( pluginList ) =>
 				pluginList.map( ( item ) => {
-					if ( item.error !== null ) {
-						return Object.assign( {}, item, { error: item.error.toString() } );
+					if ( item.error != null ) {
```

Comment:
Not 100% sure how this worked before, but I noticed we're accessing `error.code` [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plugins/jetpack-plugins-setup/index.jsx#L328) and [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plugins/jetpack-plugins-setup/index.jsx#L476) - any concerns with the new error format?

Reply:
I see: our code accesses the error properties (`code`, `message`, `name`), but the persistence code can deserialize a string value.

I changed the serialization code to save selected properties (`{ error, code, name, message }`) instead. Then the deserialized object has all the expected properties. The only difference is that it's not an `Error` instance.

By the way, I'm afraid the error handling in `jetpack-plugin-setup` is already broken. The errors returned from failed `wpcom` requests don't have a `code` field. The "error codes" like `already_registered` are stored in `error.error`. And the HTTP status code is in `error.statusCode`.

I also haven't found the `already_registered` error anywhere in Jetpack or wpcom server code. Let's ping @ryelle who worked on this (e.g., #9901) 5 years ago 🙂

Reply:
@tyxla By the way, when selecting properties from the `Error` object, I found `_.pick` indispensable. It has the nice behavior that it doesn't assign properties that are not present. For example, the straightforward vanilla JS alternative:
```js
const { name, code, message } = error;
const serialized = { name, code, message };
```
would convert `{ message: 'foo' }` to
```
{ name: undefined, code: undefined, message: 'foo' }
```

`_.pick` produces just `{ message: 'foo' }`.

Reply:
https://github.com/Automattic/wp-calypso/blob/70ac66f4e78ada28f6ad46600d54271303859b80/client/state/plugins/premium/actions.js#L284 seems to be the only place that manually mutates and assigns the `code` to be `already_registered`

Reply:
And yeah, I took a quick look at the code when reviewing the PR yesterday, and couldn't figure out how that would work, specifically when attempting to access the `code` property.

Reply:
> @tyxla By the way, when selecting properties from the `Error` object, I found `_.pick` indispensable. It has the nice behavior that it doesn't assign properties that are not present. For example, the straightforward vanilla JS alternative:
> 
> ```js
> const { name, code, message } = error;
> const serialized = { name, code, message };
> ```
> 
> would convert `{ message: 'foo' }` to
> 
> ```
> { name: undefined, code: undefined, message: 'foo' }
> ```
> 
> `_.pick` produces just `{ message: 'foo' }`.

What I don't like about it is that it's again trying to be smart about it. When specifically assigning a `name, code, message` props, I'm expecting I'll get those defined, regardless of their value. Not getting them is unexpected behavior of the code IMHO, some magic behind the scenes that could lead to subtle bugs.

Reply:
Errors are notoriously amorphous objects though. Very rarely do we know exactly what will be thrown at us. TypeScript doesn't support typing exceptions (unlike e.g. Java) at all.

Reply:
> the only place that manually mutates and assigns the code to be `already_registered`

Thanks, I didn't notice that! So the `already_registered` error is synthentic, it doesn't come from a REST response :+1:

Reply:
> Errors are notoriously amorphous objects though. Very rarely do we know exactly what will be thrown at us. TypeScript doesn't support typing exceptions (unlike e.g. Java) at all.

And that is another reason why being specific and explicit is better than relying on magic tricks IMHO.

Reply:
@tyxla I removed the `pick` usage in favor of a `map`/`filter`/`Object.fromEntries` chain that does the same thing.

I think we are already splitting hairs when arguing whether `e = {}` vs `e = { code: undefined }` makes any difference when accessing `e.code`. It's unlikely that it really matters.

I believe the error handling now works correctly, as it did before, and the PR is ready for final review.